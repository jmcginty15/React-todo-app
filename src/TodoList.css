.TodoList {
    width: 50%;
    margin: auto;
}

h1 {
    text-align: left;
}