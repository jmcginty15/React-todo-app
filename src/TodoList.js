import { useState } from 'react';
import { v4 as uuid } from 'uuid';
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import './TodoList.css';

const TodoList = () => {
    const [todos, setTodos] = useState([]);

    function removeItem(uuid) {
        const newList = [...todos];
        let i = 0;
        while (newList[i].id !== uuid && i < newList.length) i++;
        if (i !== newList.length) newList.splice(i, 1);
        setTodos(newList);
    }

    function addItem(task) {
        const newItem = { id: uuid(), task: task, complete: false}
        const newList = [...todos, newItem];
        setTodos(newList);
    }

    function editItem(uuid, newTask) {
        const newList = [...todos];
        let i = 0;
        while (newList[i].id !== uuid && i < newList.length) i++;
        if (i !== newList.length) newList[i].task = newTask;
        setTodos(newList);
    }

    return (
        <div className="TodoList">
            <h1>Todo:</h1>
            {todos.map(todo => <Todo
                key={todo.id}
                id={todo.id}
                task={todo.task}
                complete={todo.complete}
                handleRemove={(id) => removeItem(id)}
                handleEdit={(id, task) => editItem(id, task)}
            />)}
            <NewTodoForm addItem={addItem} />
        </div>
    )
}

export default TodoList;