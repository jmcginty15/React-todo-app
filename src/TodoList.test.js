import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

// smoke test
it('renders without crashing', function () {
    render(<TodoList />);
});

// snapshot test
it('matches snapshot', function () {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it('adds a new item on form submit', function () {
    const { queryByText, queryByTestId } = render(<TodoList />);
    const inputField = queryByTestId('new-todo-input');
    const submitButton = queryByText('Add');
    fireEvent.change(inputField, { target: { value: 'feed dog' } });
    fireEvent.click(submitButton);

    const newItem = queryByText('feed dog');
    expect(newItem).toBeInTheDocument();
});

it('removes an item on button click', function () {
    const { queryByText, queryByTestId } = render(<TodoList />);
    const inputField = queryByTestId('new-todo-input');
    const submitButton = queryByText('Add');
    fireEvent.change(inputField, { target: { value: 'feed dog' } });
    fireEvent.click(submitButton);

    const newItem = queryByText('feed dog');
    expect(newItem).toBeInTheDocument();

    const removeButton = queryByText('X');
    fireEvent.click(removeButton);
    expect(newItem).not.toBeInTheDocument();
});

it('marks an item complete or incomplete', function () {
    const { queryByText, queryByTestId } = render(<TodoList />);
    const inputField = queryByTestId('new-todo-input');
    const submitButton = queryByText('Add');
    fireEvent.change(inputField, { target: { value: 'feed dog' } });
    fireEvent.click(submitButton);

    const markButton = queryByText('Mark Complete');
    fireEvent.click(markButton);
    const todoText = queryByTestId('todo-text');
    expect(queryByText('Mark Complete')).not.toBeInTheDocument();
    expect(queryByText('Mark Incomplete')).toBeInTheDocument();
    expect(todoText.className).toBe('Todo-complete');

    fireEvent.click(markButton);
    expect(queryByText('Mark Complete')).toBeInTheDocument();
    expect(queryByText('Mark Incomplete')).not.toBeInTheDocument();
    expect(todoText.className).toBe('');
});

it('edits an item through the edit form', function () {
    const { queryByText, queryByTestId } = render(<TodoList />);
    const inputField = queryByTestId('new-todo-input');
    const submitButton = queryByText('Add');
    fireEvent.change(inputField, { target: { value: 'feed dog' } });
    fireEvent.click(submitButton);

    const editButton = queryByText('Edit');
    fireEvent.click(editButton);
    const editInputField = queryByTestId('edit-todo-input');
    const editSubmitButton = queryByText('Submit');
    expect(editInputField).toBeInTheDocument();
    expect(editSubmitButton).toBeInTheDocument();
    fireEvent.change(editInputField, { target: { value: 'feed cat' } });
    fireEvent.click(editSubmitButton);
    expect(queryByText('feed dog')).not.toBeInTheDocument();
    expect(queryByText('feed cat')).toBeInTheDocument();
});

it('shows original item when edit is canceled', function () {
    const { queryByText, queryByTestId } = render(<TodoList />);
    const inputField = queryByTestId('new-todo-input');
    const submitButton = queryByText('Add');
    fireEvent.change(inputField, { target: { value: 'feed dog' } });
    fireEvent.click(submitButton);

    const editButton = queryByText('Edit');
    fireEvent.click(editButton);
    const editCancelButton = queryByText('Cancel');
    fireEvent.click(editCancelButton);
    expect(queryByText('feed dog')).toBeInTheDocument();
});