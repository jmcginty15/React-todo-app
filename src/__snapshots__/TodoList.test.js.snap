// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <h1>
      Todo:
    </h1>
    <div
      class="NewTodoForm"
    >
      <form>
        <input
          data-testid="new-todo-input"
          name="task"
          placeholder="enter a new todo item"
          required=""
          type="text"
          value=""
        />
        <button
          class="NewTodoForm-submit"
        >
          Add
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
